### PR TITLE
Introduction of UNKNOWN state

### LINKED ISSUE #647: Introduce  "UNKNOWN" job state as described here:
@p1003 what about the "UNKNOWN" state as described here:

> In case the notification system (watcher.py) is broken and in the database we have no information at all, then all the tasks and simulations in such state should have a new state, called UNKNOWN. This may be implemented easily on creation of the objects in the DB, instead of creating the object in PENDING state, give them state UNKNOWN.
> 
> For Celery transition from UNKNOWN to PENDING happens when simulation will get job_id from Celery.
> For SLURM this transition happens when we obtain job id from slurm. Here it may be time consuming as to get job id we need to call slurm via fabric/ssh.
> Simulations with UNKNOWN state will have empty list of tasks. This should be properly displayed in the UI.
> 
> When listing simulations we should list simulation with UNKNOWN state as well.

_Originally posted by @grzanka in https://github.com/yaptide/yaptide/issues/529#issuecomment-1743048294_

### PR DESCRIPTION
**Pull request by p1003**

### CODE REVIEW

**Comment by grzanka** [comment]
Those are the states:
```
class EntityState(Enum):
    """Job state types - move it to more utils like place in future"""

    UNKNOWN = "UNKNOWN"
    PENDING = "PENDING"
    RUNNING = "RUNNING"
    CANCELED = "CANCELED"
    COMPLETED = "COMPLETED"
    FAILED = "FAILED"
```


Is this a correct state graph ?

```mermaid
---
title: Job states
---
stateDiagram-v2
    [*] --> UNKNOWN

    UNKNOWN --> PENDING
    PENDING --> RUNNING
    RUNNING --> COMPLETED
    COMPLETED --> [*]

    RUNNING --> FAILED
    FAILED --> [*]

    PENDING --> CANCELED
    RUNNING --> CANCELED
    CANCELED --> [*]

    UNKNOWN --> CANCELED

```

**Comment by grzanka** [comment]
Is it still draft ?

**Comment by grzanka** [comment]
@p1003 can you discuss with @Derstilon and @ostatni5 how this could be reflected in UI ?

**Comment by p1003** [comment]
@grzanka diagram looks fine but canceling job with `UNKNOWN` state might cause some problems for now. As a quick fix I can forbid canceling jobs in this state, because it might cause some problems:
- if job has `UNKNOWN` state, that might mean the submition has not yet been finished
- Celery tasks need to send their internal ids to database first, because otherwise they cannot be canceled - fix for this might include checking if task is canceled after it sends it's update and sending appropriate message to it, so that he can cancel itself

Currently server sends responses on POST methods submitting jobs, after the simulations' states are changed to `PENDING`. This might change when we start working on this issue [https://github.com/yaptide/yaptide/issues/532](url).

**Comment by grzanka** [comment]
> @grzanka diagram looks fine but canceling job with `UNKNOWN` state might cause some problems for now. As a quick fix I can forbid canceling jobs in this state, because it might cause some problems:
> - if job has `UNKNOWN` state, that might mean the submition has not yet been finished
> - Celery tasks need to send their internal ids to database first, because otherwise they cannot be canceled - fix for this might include checking if task is canceled after it sends it's update and sending appropriate message to it, so that he can cancel itself
> 
> Currently server sends responses on POST methods submitting jobs, after the simulations' states are changed to `PENDING`. This might change when we start working on this issue [https://github.com/yaptide/yaptide/issues/532](url).

Ok, then suggest updated diagram.

**Comment by grzanka** [comment]
Can you add as well a single markdown file in docs folder with state documentation ?

**Comment by p1003** [comment]
```mermaid
---
title: Job states
---
stateDiagram-v2
    [*] --> UNKNOWN

    UNKNOWN --> PENDING
    PENDING --> RUNNING
    RUNNING --> COMPLETED
    COMPLETED --> [*]

    RUNNING --> FAILED
    FAILED --> [*]

    PENDING --> CANCELED
    RUNNING --> CANCELED
    CANCELED --> [*]

```

File: yaptide/utils/enums.py [hunk 1]
```diff
@@ -11,6 +11,7 @@ class PlatformType(Enum):
 class EntityState(Enum):
     """Job state types - move it to more utils like place in future"""
```
Comment (grzanka) [inline yaptide/utils/enums.py]:
```suggestion
    """Job state types"""
```

File: yaptide/routes/utils/utils.py [hunk 2]
```diff
@@ -11,3 +15,32 @@ def check_if_job_is_owned_and_exist(job_id: str, user: UserModel) -> tuple[bool,
     if simulation.user_id == user.id:
         return True, "", 200
     return False, 'Job with provided ID does not belong to the user', 403
+
+
+def determine_input_type(payload_dict: dict) -> Union[str, None]:
```
Comment (grzanka) [inline yaptide/routes/utils/utils.py]:
```suggestion
def determine_input_type(payload_dict: dict) -> Optional[str]:
```

File: yaptide/routes/utils/utils.py [hunk 3]
```diff
@@ -1,5 +1,9 @@
+from typing import Union
```
Comment (grzanka) [inline yaptide/routes/utils/utils.py]:
```suggestion
from typing import Optional
```

**Review by grzanka — APPROVED** [review]

**Pull request merged**